### PR TITLE
chore(application-system): using aws-sdkv3 initial implementation

### PR DESCRIPTION
**Pull request by HjorturJ**
# ...

https://app.asana.com/0/1207402031871459/1208111779122042/f

## What

Implement s3 aws sdk v3 without breaking anything

## Why
because aws sdk v2 will be depricated in 2025

## Screenshots / Gifs

Attach Screenshots / Gifs to help reviewers understand the scope of the pull request

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Release Notes

- **New Features**
  - Enhanced file upload and export functionalities with improved parameter structures for AWS S3 operations.
  - Introduced new methods for file handling in the AWS service, including presigned URL generation.

- **Bug Fixes**
  - Improved error handling for PDF generation to enhance reliability.

- **Documentation**
  - Updated test suite for AWS service to reflect new method names and functionalities.

### CODE REVIEW

File: libs/nest/aws/src/lib/aws.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+      const startingString = 'Hello ÁáÉéÚúÍíÓóÐðÞþ'
+      const expectedResult = Buffer.from(startingString).toString(
+        encoding as BufferEncoding,
+      )
+
+      mockGetObjectCommandWithBodyOnce(startingString)
+      const result = await awsService.getFileContent(
+        { bucket: 'abc', key: 'def' },
+        encoding as EncodingString,
+      )
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.spec.ts:79]: Cant easily do this as it conflicts with how we pass the params into the function through .each() and the fact that BufferEncoding and EncodingString arent the same type

File: libs/application/template-api-modules/src/lib/modules/templates/parental-leave/parental-leave.service.ts [hunk 2]
```diff
@@ -385,16 +383,19 @@ export class ParentalLeaveService extends BaseTemplateApiService {
...
+        },
+        'base64',
+      )
 
       if (!fileContent) {
         throw new Error('File content was undefined')
       }
 
-      return fileContent.toString('base64')
+      return fileContent
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/application/template-api-modules/src/lib/modules/templates/parental-leave/parental-leave.service.ts:398]: It is the same return type, a string containing the file content

File: libs/nest/aws/src/lib/aws.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+})
+
+const mockGetObjectCommandWithBodyOnce = (body: string) => {
+  const stream = new Readable()
+  stream.push(body)
+  stream.push(null) // end of stream
+  const sdkStream = sdkStreamMixin(stream)
+
+  s3Mock.on(GetObjectCommand).resolvesOnce({ Body: sdkStream })
+}
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.spec.ts:172]: Hmm possibly a good idea, but I dont see many other test classes mocking the actual s3 client, they would instead mock the awsService functions.

File: libs/nest/aws/src/lib/aws.service.spec.ts [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+      expect(result).toEqual(expectedResult)
+    },
+  )
+
+  it('should return a valid presigned url', async () => {
+    const expectedResult = 'https://website.com'
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const getSignedUrlMock: jest.Mock = getSignedUrl as any
+    getSignedUrlMock.mockResolvedValue(expectedResult)
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.spec.ts:89]: Changed.

File: libs/nest/aws/src/lib/aws.service.spec.ts [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+  it('should return a file from a bucket if one exists', async () => {
+    const expectedResult = 'hello world'
+
+    mockGetObjectCommandWithBodyOnce(expectedResult)
+
+    const getObjectResult = await awsService.getFile({ bucket: 'x', key: 'y' })
+    if (!getObjectResult) fail('getObjectResult was undefined')
+
+    const contentResults = await getObjectResult.Body?.transformToString()
+    if (!contentResults) fail('raw transform result was undefined or empty')
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.spec.ts:54]: Fixed.

File: libs/nest/aws/src/lib/aws.service.ts [hunk 6]
```diff
@@ -3,72 +1,174 @@
...
+        return false
+      } else {
+        // some other error
+        this.logger.error(
+          'Error occurred while checking if file exists in S3',
+          error,
+        )
+        return false
+      }
+    }
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts:125]: Fixed.

File: libs/nest/aws/src/lib/aws.service.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.spec.ts:54]: Fixed.

File: libs/nest/aws/src/lib/aws.service.ts [hunk 7]
```diff
@@ -3,72 +1,174 @@
...
+  }
+
+  public getBucketKey(BucketKeyPairOrFilename: BucketKeyPair | string): {
+    bucket: string
+    key: string
+  } {
+    return typeof BucketKeyPairOrFilename === 'object'
+      ? BucketKeyPairOrFilename
+      : AmazonS3URI(BucketKeyPairOrFilename)
+  }
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts:156]: Fixed.

File: libs/nest/aws/src/lib/aws.service.ts [hunk 8]
```diff
@@ -3,72 +1,174 @@
...
+      if (
+        result.$metadata.httpStatusCode !== 204 &&
+        result.$metadata.httpStatusCode !== 200
+      ) {
+        throw new Error('Unexpected http response when deleting object from S3')
+      }
+    } catch (error) {
+      this.logger.error('Error occurred while deleteing file from S3', error)
+    }
+  }
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts:147]: Sure, fixed.

File: libs/nest/aws/src/lib/aws.service.ts [hunk 9]
```diff
@@ -1,74 +1,174 @@
...
+      )
+
+      if (
+        result.$metadata.httpStatusCode !== 204 &&
+        result.$metadata.httpStatusCode !== 200
+      ) {
+        throw new Error('Unexpected http response when deleting object from S3')
+      }
+    } catch (error) {
+      this.logger.error('Error occurred while deleteing file from S3', error)
```
[the remark this answers is not in the document]
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.service.ts:145]: Fixed.

**Review by AndesKrrrrrrrrrrr — APPROVED** [review]

**Review by norda-gunni — APPROVED** [review]
LGTM

**Review by birkirkristmunds — APPROVED** [review]
LGTM!

**Review by rafnarnason — APPROVED** [review]

File: libs/nest/aws/src/lib/aws.module.ts [hunk 10]
```diff
@@ -1,9 +1,16 @@
...
-  providers: [AwsService],
-  exports: [AwsService],
+  providers: [
+    AwsService,
+    {
+      provide: S3Client,
+      useValue: new S3Client(),
+    },
+  ],
+  exports: [AwsService, S3Client],
```
Comment (eirikurn) [inline libs/nest/aws/src/lib/aws.module.ts:14]: Do we need/want to export the S3Client in addition to the AwsService (which might be more appropriately named S3Service)?

**Review by eirikurn — APPROVED** [review]
Approving core files with one question. Great job!

File: libs/nest/aws/src/lib/aws.module.ts (hunk 10 above)
Comment (HjorturJ) [reply libs/nest/aws/src/lib/aws.module.ts]: Whoops, no we dont, thought I had removed that, fixed now.
Promise to rename the service to S3Service in the next PR when I start placing it all over the place 😅

**Pull request merged**